File: service1/src/main/java/ru/ifmo/soa/service1/persons/repository/PersonRepository.java
package ru.ifmo.soa.service1.persons.repository;

import jakarta.enterprise.context.RequestScoped;
import jakarta.inject.Inject;
import ru.ifmo.soa.service1.app.db.DBConnectionManager;
import ru.ifmo.soa.service1.persons.model.Country;
import ru.ifmo.soa.service1.persons.model.Person;

import java.sql.*;
import java.util.Optional;

@RequestScoped
public class PersonRepository {

    @Inject
    DBConnectionManager connectionManager;

    public void delete(Person person) throws SQLException {
        if (person.getPassportID() == null) return;

        String sql = "DELETE FROM Persons WHERE passport_id = ?;";
        Connection connection = connectionManager.getConnection();
        PreparedStatement statement = connection.prepareStatement(sql);
        statement.setString(1, person.getPassportID());
        statement.execute();
        connection.close();

    }

    public void create(Person person) throws SQLException {

        String sql = "INSERT INTO " +
                "Persons (passport_id, name, height, weight, nationality)" +
                "VALUES (?, ?, ?, ?, ?);";
        Connection connection = connectionManager.getConnection();
        PreparedStatement statement = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
        statement.setString(1, person.getPassportID());
        statement.setString(2, person.getName());
        statement.setLong(3, person.getHeight());
        statement.setDouble(4, person.getWeight());
        statement.setString(5, person.getNationality().toString());
        statement.execute();
        connection.close();
    }

    public void update(Person person) throws SQLException {

        String sql = "UPDATE persons SET name = ?, height = ?, weight = ?, nationality = ?" +
                "WHERE passport_id = ?;";
        Connection connection = connectionManager.getConnection();
        PreparedStatement statement = connection.prepareStatement(sql);
        statement.setString(1, person.getName());
        statement.setLong(2, person.getHeight());
        statement.setDouble(3, person.getWeight());
        statement.setString(4, person.getNationality().toString());
        statement.setString(5, person.getPassportID());
        statement.execute();
        connection.close();
    }

    public Optional<Person> getById(String passportId) throws SQLException {
        String sql = "SELECT * FROM Persons  WHERE passport_id = ?";
        Connection connection = connectionManager.getConnection();
        PreparedStatement statement = connection.prepareStatement(sql);
        statement.setString(1, passportId);
        statement.execute();
        ResultSet resultSet = statement.getResultSet();
        Person person = null;
        if (resultSet.next()) person = fromRow(resultSet);
        return Optional.ofNullable(person);
    }

    Person fromRow(ResultSet resultSet) throws SQLException {
        Person.PersonBuilder personBuilder = Person.builder();
        personBuilder
                .passportID(resultSet.getString("passport_id"))
                .name(resultSet.getString("name"))
                .height(resultSet.getLong("height"))
                .weight(resultSet.getDouble("weight"))
                .nationality(Country.valueOf(resultSet.getString("nationality")));
        return personBuilder.build();

    }


}

File: frontend/static/data/repository/person_repository.js
import {Person} from "../model/person.js";
import {request_crud} from "../../api/api.js";

class PersonRepository {

    getPerson(id, callback) {
        request_crud(
            `persons/${id}/`,
            {},
            'GET',
            (data) => {
                let item = data['person']
                let person = new Person(
                    item['name'],
                    item['height'],
                    item['weight'],
                    item['passportID'],
                    item['nationality']
                )
                callback(person)
            },
            {},
            (error) => callback(null, error)
        )
    }

    createPerson(fields, callback) {
        request_crud(
            'persons/',
            {
                'person': JSON.stringify({
                    'name': fields['name'],
                    'height': fields['height'],
                    'weight': fields['weight'],
                    'passportID': fields['passportID'],
                    'nationality': fields['nationality']
                })
            },
            'POST',
            (_) => callback(true),
            {},
            (error) => callback(null, error)
        )
    }

    updatePerson(id, fields, callback) {
        request_crud(
            `persons/${id}/`,
            {
                'person': JSON.stringify({
                    'name': fields['name'],
                    'height': fields['height'],
                    'weight': fields['weight'],
                    'nationality': fields['nationality']
                })
            },
            'PUT',
            (_) => callback(true),
            {},
            (error) => callback(null, error)
        )
    }

    deletePerson(id, callback) {
        request_crud(
            `persons/${id}/`,
            {},
            'DELETE',
            (_) => callback(true),
            {},
            (error) => callback(null, error)
        )
    }
}

export let personRepository = new PersonRepository()
